src/include/miopen/graphapi/enginecfg.hpp

```diff
@@ -0,0 +1,118 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2024 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#pragma once
+
+#include <miopen/graphapi/engine.hpp>
+
+namespace miopen {
+
+namespace graphapi {
+
+class EngineCfg
+{
+private:
+    /* we don't use a pointer here to allow a user
+     * to have several configs for an Engine. Each
+     * config might modify its Engine in future so
+     * their instance of Engine shouldn't be shared
+     */
+    Engine mEngine;
```

Comment:
I'm not sure I agree with this. I think this should be a pointer. Backend API can set engine with a pointer, but also call `getAttribute` on the same engine, so we should be returning the same engine pointer. 
Further, a user may modify a `knob` via the `EngineConfig` but read it via the `Engine` (bizarre API choice IMO), but to support it means we keep a pointer.

Reply:
I thought the instance of Engine provided by a user should not be modified by setting a knob and reading the same knob from the Engine should return the default value; and EngineCfg represents an engine's config without touching the source engine instance.

But the behavior is not described in detail in the docs so you may be right.

Reply:
This engine descriptor and its way of 'creating' it by index seems to purpose the case when a user wants to go through available engines and try different knob values to find a suitable engine/its config. Then the user can save found combination by serialization and in future reuse it without creating all the needed descriptors and providing only variant pack with uids/data pointers (even no need of tensor descriptors).

If a user wants the lib to find the best engine/its config, there is engineheir descriptor.

Reply:
engineheir descriptor returns list of enginecfg descriptors skipping 'engine' step.